### PR TITLE
Add initial code for static block game agents

### PR DESCRIPTION
**Pull request by ethanp55**

### CODE REVIEW

File: game/block_game.py [hunk 1]
```diff
@@ -11,19 +11,28 @@
...
 class BlockGameState(State):
     ''' Abstract State class '''
 
     CENTER = 3
+
     def __init__(self):
         self.blocks = [3 for _ in range(0, 9)]
         self.turn = P1
 
+    def get_round_num(self):
```
Comment (TimWhiting) [inline game/block_game.py]: ply num? It's not really a round

File: game/block_game_agents_example.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+# -------------------------------------------------------------------
+# ------------------------ STATIC AGENTS ----------------------------
+# -------------------------------------------------------------------
+def minimax_func(reward_action_pairs: List[Tuple[Tuple[float, float], int]], turn: int) -> Tuple[Tuple[float, float], int]:
+    other_player = P1 if turn == P2 else P2
+    return min(reward_action_pairs, key=lambda x: x[0][other_player])
+
+
+minimax_agent = FixedPolicyBlockGameAgent(
+    minimax_func, 'Minimax', deepcopy(block_game_tree))
```
Comment (TimWhiting) [inline game/block_game_agents_example.py]: Is the deepcopy needed? The agents shouldn't alter the game tree, just store their own desired actions for the game. Oh, I see you are storing ideal_reward in the tree node. Maybe store that in the agent instead so deepcopy is not needed?

File: game/block_game_tree.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        queue.append(head_node)
+        head_node.visited = True
+        head_node.state_id = state_id
+        self.state_id_map[state_id] = head_node
+
+        while queue:
+            # Dequeue a node from the queue
+            new_node = queue.pop(0)
+
+            # Get all adjacent nodes of the dequeued node new_node. If an adjacent node has not been visited, mark and enqueue it
```
Comment (TimWhiting) [inline game/block_game_tree.py]: This isn't adjacent nodes of new_node, this is children right?

**Review by TimWhiting — CHANGES_REQUESTED** [review]
Can you also add *.pyc to the gitignore?

File: game/block_game_tree.py (hunk 3 above)
Comment (ethanp55) [reply game/block_game_tree.py]: Yea children, that's what I meant

File: game/block_game.py (hunk 1 above)
Comment (ethanp55) [reply game/block_game.py]: K, I'll change the name

File: game/block_game_tree.py (hunk 3 above)
Comment (ethanp55) [reply game/block_game_tree.py]: I'll update the comment so it's clearer

File: game/block_game_agents_example.py (hunk 2 above)
Comment (ethanp55) [reply game/block_game_agents_example.py]: Alright, 1 sec

Comment (ethanp55) [reply game/block_game_agents_example.py]: Done

File: game/block_game_agent.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        for tree_node in state_map.values():
+            if tree_node.state.is_terminal():
+                p1_reward = tree_node.state.reward(P1)
+                p2_reward = tree_node.state.reward(P2)
+
+                ideal_reward_map[tree_node] = (p1_reward, p2_reward)
+                self.policy[str(tree_node.state)] = ACTIONS[-1]
+
+            else:
+                reward_action_pairs = [(ideal_reward_map[tree_node.action_to_children_map[i]], i)
```
Comment (TimWhiting) [inline game/block_game_agent.py:44]: No need to change, but is there a reason why you chose to do reward: action rather than action: reward pairs?

**Pull request merged**